### ISSUE TITLE
Allow for user to create a node with attached disks

### ISSUE THREAD

**Issue by kannon92** [metal3-io/cluster-api-provider-metal3#1910]
**User Story**

As an operator, I would like to create a node with a root partition and attached disks.

The API should be optional and allow for an array of disks in case a user wants to add more than 1 disk to the node.

**Detailed Description**

AWS, GCP and Azure CAPI providers allow one to create a node with additional disks. It would be great for this provider to also allow this functionality.

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature

**Comment by kannon92** [comment]
The main use cases we have are when a user booted a node and then they realize that they want to add a disk rather than recreate the node.

A few usecases I have in mind would be supporting https://kubernetes.io/blog/2024/01/23/kubernetes-separate-image-filesystem/.

Generally we find users realize that they are running out of space on their cluster. So they would resort to adding a new disk to a node to help alleviate the situation.

Another one that we would be interested to support is https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200423-etcd-data-disk.md.

Adding etcd to a separate disk.

**Comment by jayofdoom** [comment]
@kannon92 is an old co-worker of mine and asked me to take a look. My concerns line up mostly with Dmitry, we need to figure out "attach" means to you in context of a physical machine -- what should be the other side of that "attach" API. Seems to me in order for "attach" to make sense, you kinda need a source (what do I attach?) and a destination (where do I attach it?).

If we're talking physical disks; those cannot generally be dynamically added/removed from a machine via an API.
If we're talking network-attached disks, you'd likely need some coordination with metal3/ironic around when the node reboots (Ironic handles this via our existing StorageInterface; which AFAICT doesn't have native support for any kubernetes-flavored way of storage management -- if such a thing exists).

**Comment by dtantsur** [comment]
> If we're talking physical disks; those cannot generally be dynamically added/removed from a machine via an API.

Yeah, except for very exotic composable hardware (and even it will probably require a reboot). I'm still not sure what can be done from the *hardware* perspective.

**Comment by dtantsur** [comment]
Could you please provide further clarifications? We'd like to know what you mean under "attach" exactly, and why it cannot be done by a software component triggered on the first boot (cloud-init/ignition).

**Comment by kannon92** [comment]
> Could you please provide further clarifications? We'd like to know what you mean under "attach" exactly, and why it cannot be done by a software component triggered on the first boot (cloud-init/ignition).

I think the goal would be shift cluster to be more declarative. Get me a node with this filesystem layout and install stuff on it.

Yes, you can do most of this with first boot but that is hard to optimize across many clusters. 

I just mean "attach" to say I want to create a node with a particular disk setup. 

The main use cases I elaborated in the original thread (swap nodes,  container runtime filesystems, etc). You can do most of this with ignition but its mostly a node as a pet analogy.

**Comment by Rozzii** [comment]
It would be possible to do partitioning in the preprovisionig/provisioning phase maybe even building overlay file systems and similar things but 

"a node with a particular disk setup" is very broad requirement, do you need to configure block storage e.g SCSI, FCOE, multipath for SCSI multipath for FCOE, encryption, NFS,  Ceph  block storage , Ceph NFS  some other proprietary storage solution? 

Generally we try to stay away from meddling in the config of a provisioned node that is why cloud-init/ignition or similar config manager is the recommended way and that is why it is possible to inject node specific cloud-init network,meta and user data files with BMO and that is why CAPM3 has the has the metal3datatemplate resource.
